### PR TITLE
Postgres, jsonschema

### PR DESCRIPTION
**Pull request by mirceachira**

### CODE REVIEW

File: scrape_rec/utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import uuid
+
+import sqlalchemy as sa
+from sqlalchemy.ext.declarative import declarative_base  
+
+
+POSTGRES_USER = 'postgres'
```
Comment (danielacraciun) [inline scrape_rec/utils.py]: These should be private. Fetch them from env?

File: scrape_rec/spiders/base_realestate.py [hunk 2]
```diff
@@ -20,14 +22,16 @@ def process_ad_date(self, ad_date):
         return ad_date
 
     def process_price(self, price):
-        return price
+        return price, 'EUR'
```
Comment (danielacraciun) [inline scrape_rec/spiders/base_realestate.py]: :+1:

File: scrape_rec/spiders/base_realestate.py [hunk 3]
```diff
@@ -65,4 +69,4 @@ def parse(self, response):
             yield response.follow(link, callback=self.process_link)
 
         next_link = response.xpath(self.next_link_xpath).extract_first()
-        yield response.follow(next_link)
\ No newline at end of file
+        yield response.follow(next_link, dont_filter=True)
```
Comment (danielacraciun) [inline scrape_rec/spiders/base_realestate.py]: What's the meaning of `dont_filter` here?

File: scrape_rec/settings.py [hunk 4]
```diff
@@ -1,94 +1,30 @@
...
-# See also autothrottle settings and docs
-DOWNLOAD_DELAY = 3
-
-# The download delay setting will honor only one of:
-#CONCURRENT_REQUESTS_PER_DOMAIN = 16
-#CONCURRENT_REQUESTS_PER_IP = 16
-
-# Disable cookies (enabled by default)
-#COOKIES_ENABLED = False
+DOWNLOAD_DELAY = 1
```
Comment (danielacraciun) [inline scrape_rec/settings.py]: What's this do?

File: scrape_rec/pipelines.py [hunk 5]
```diff
@@ -53,4 +48,37 @@ def process_item(self, item, spider):
...
+
+class PostgresPipeline(object):
+
+    def __init__(self):
+        self.session = get_postgres_session()
+
+    def process_item(self, item, spider):
+        # TODO: don't add if already exists for double check in case redis fails
+
+        entry = RealestateApartment(
```
Comment (danielacraciun) [inline scrape_rec/pipelines.py]: Why not
```suggestion
        entry = RealestateApartment(**item)
```
?

File: scrape_rec/items.py [hunk 6]
```diff
@@ -1,27 +1,87 @@
...
+            'cellar': {
+                'description': 'Has or not cellar',
+                'type': 'boolean'
+            },
+            'source_website': {
+                'description': 'The website of origin',
+                'type': 'string'
+            },
+            'source_offer': {
+                'description': 'Is or not sold by an agency',
```
Comment (danielacraciun) [inline scrape_rec/items.py]: Update description

File: scrape_rec/items.py [hunk 7]
```diff
@@ -1,27 +1,87 @@
...
+                'minimum': 0
+            },
+            'building_year': {
+                'description': 'The building construition date',
+                'type': 'string'
+            },
+            'floor': {
+                'description': 'Obvious',
+                'type': 'integer',
+                'minimum': 0
```
Comment (danielacraciun) [inline scrape_rec/items.py]: Has no minimum, could be basement. (Cluj rents amirite)

File: scrape_rec/items.py [hunk 8]
```diff
@@ -1,27 +1,87 @@
...
+                'description': 'Apartment sections',
+                'type': 'string'
+            },
+            'surface': {
+                'description': 'Surface in mp^2',
+                'type': 'integer',
+                'minimum': 0
+            },
+            'building_year': {
+                'description': 'The building construition date',
```
Comment (danielacraciun) [inline scrape_rec/items.py]:
```suggestion
                'description': 'The building construction date',
```

File: scrape_rec/items.py [hunk 9]
```diff
@@ -1,27 +1,87 @@
...
+                'type': 'string'
+            },
+            'floor': {
+                'description': 'Obvious',
```
Comment (danielacraciun) [inline scrape_rec/items.py]: V helpful.

File: scrape_rec/items.py [hunk 10]
```diff
@@ -1,27 +1,87 @@
...
+            },
+            'price': {
+                'type': 'number',
+                'minimum': 0,
+                'exclusiveMinimum': True
+            },
+            'currency': {
+                'description': 'The creation date of the listing or if not found the scrape date',
+                'type': 'string',
+                # 'pattern': 'EUR|RON|USD'
```
Comment (danielacraciun) [inline scrape_rec/items.py]: Bring back the patterns

File: scrape_rec/spiders/base_realestate.py (hunk 3 above)
Comment (mirceachira) [reply scrape_rec/spiders/base_realestate.py]: No longer relevant but scrapy has a default duplicate filter that doesn't let you go to the same url twice. In our case, the idea was to use a redis filter but checking postgres pk is better

File: scrape_rec/settings.py (hunk 4 above)
Comment (mirceachira) [reply scrape_rec/settings.py]: This is the time to wait between requests. In the back scrapy also has a setting for randomize_download_delay which ensures that the time between requests is download_delay * 0.5 <= delay <= download_delay * 1.5 so you have some variation and it's harder to get detected

File: scrape_rec/items.py (hunk 10 above)
Comment (mirceachira) [reply scrape_rec/items.py]: Fixed

File: scrape_rec/items.py (hunk 9 above)
Comment (mirceachira) [reply scrape_rec/items.py]: Thank you

File: scrape_rec/items.py (hunk 8 above)
Comment (mirceachira) [reply scrape_rec/items.py]: Fixed

File: scrape_rec/utils.py (hunk 1 above)
Comment (danielacraciun) [reply scrape_rec/utils.py]: Let's do this later

**Comment by danielacraciun** [comment]
Should I merge this? @mirceachira

**Pull request merged**